File: backend/games/random_checkout.py
"""random_checkout-Familie (Random Checkout). Siehe SPEC §25 und
docs/ARCHITEKTUR.md Abschnitt 4.

Korrektur (08.09.2026): Ein Versuch ist GENAU EINE Aufnahme (3 Darts) -
kein mehrteiliger Versuch mehr, egal ob geschafft oder nicht (auch bei
Bust wechselt sofort der Spieler). Die geteilte Zufallszahl pro Runde
erzeugt die MatchEngine (Fairness, SPEC §8) - diese Familie kuemmert
sich nur um die Countdown-Logik der einzelnen Aufnahme.
"""
from __future__ import annotations

from backend.engine.scoring import apply_countdown_throw


def create_player_state(target: int) -> dict:
    return {"remaining": target, "successfulCheckouts": 0, "attempts": 0}


def apply_throw(player_state: dict, visit_throws: list[dict], settings: dict) -> dict:
    # Double-Out ist beim Checkout-Training immer an - kein eigener
    # Schalter in den Settings (SPEC §25 nennt keinen).
    return apply_countdown_throw(player_state["remaining"], visit_throws, True)

File: frontend/src/engine/families/randomCheckout.ts
// 1:1-Port von backend/games/random_checkout.py (Phase D des Client-
// Rewrites, siehe ~/.claude/plans/agile-brewing-wadler.md). Alle
// Spieler versuchen denselben (geteilten, fairen) Zufalls-Checkout-
// Wert - TASK_BASED_FAMILIES, siehe matchEngine.ts.
import { applyCountdownThrow, CountdownResult, Segment } from "../scoring";

export type RandomCheckoutPlayerState = {
  remaining: number;
  successfulCheckouts: number;
  attempts: number;
  attemptRemaining: number;
  taskVisitsUsed: number;
  taskDone: boolean;
};

export function createPlayerState(target: number): RandomCheckoutPlayerState {
  return {
    remaining: target,
    successfulCheckouts: 0,
    attempts: 0,
    attemptRemaining: target,
    taskVisitsUsed: 0,
    taskDone: false,
  };
}

// Double-Out ist beim Checkout-Training immer an - kein eigener
// Schalter in den Settings. Basis ist der ueber mehrere eigene
// Aufnahmen hinweg mitgefuehrte Rest des laufenden Versuchs, nicht
// der nominale (geteilte) Zielwert.
export function applyThrow(playerState: RandomCheckoutPlayerState, visitThrows: Segment[]): CountdownResult {
  return applyCountdownThrow(playerState.attemptRemaining, visitThrows, "double_out");
}

// Wird von der Engine GENAU EINMAL pro Spieler und Versuch aufgerufen.
export function resolveAttempt(playerState: RandomCheckoutPlayerState, _settings: Record<string, unknown>, success: boolean): void {
  playerState.attempts += 1;
  if (success) playerState.successfulCheckouts += 1;
}
